// react-container-component/lib/Container.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propUtils = require('./propUtils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Container = function (_React$Component) {
  _inherits(Container, _React$Component);

  function Container() {
    _classCallCheck(this, Container);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  // This is the component that my container is going to render
  Container.prototype.setComponent = function setComponent(component) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    this.childComponent = component;
    this.options = options;
    var propTypes = component.propTypes || options.propTypes;

    // when you inject dinamically a different child component using props and
    // you want to add propTypes that are not set at design time in the child component
    if (options.props && options.props._component && typeof options.props._component.propTypes) {
      propTypes = Object.assign({}, propTypes, options.props._component.propTypes);
    }

    if (typeof process !== "undefined" && process.env.NODE_ENV !== 'production') {
      if (!propTypes) {
        (0, _propUtils.logWarning)('You are rendering ' + (component.name || 'a component') + ' that has no propTypes. PropTypes are used by the Container\n        to set the props the container will pass to the child component. If ' + (component.name || 'your component') + ' doesn\'t use any props then ignore this warning');
      }
    }

    // options.addProps are props that you want to explicitly inject to the container's child component.
    // This is not needed if you name the methods or the this.state keys that you want to pass down with the same name as the child component's propTypes.
    // options.addProps is not needed for properties that are injected already in the container, for instance
    // by other functions like react-redux connect function. In most cases you don't have to use options.addProps
    var addProps = options.addProps || {};

    var src = 'return {';
    for (var key in propTypes) {
      var addProp = addProps[key];
      if (addProp && typeof addProp === 'string') {
        src += key + ': ' + addProp + ',';
      } else if (addProp) {
        (0, _propUtils.logWarning)('{ addProps: {' + key + '} must be a string, example: \'this.state.user.id\'');
      } else {
        src += key + ': this[\'' + key + '\'] || props[\'' + key + '\'] || (this.state && this.state[\'' + key + '\']),';
      }
    }

    if (propTypes) {
      src = src.slice(0, -1);
    }
    src += '}';
    this.getChildProps = Function('props', src);

    if (options.props && options.props._methods) {
      for (var _key in options.props._methods) {
        var injectMethod = options.props._methods[_key];
        if (this.hasOwnProperty(_key) && typeof this[_key] === 'function' && typeof injectMethod === 'function') {
          this[_key] = injectMethod;
        }
      }
    }
  };

  Container.prototype.render = function render() {
    // The component that will be rendered can be set in two different ways:
    // 1. Via Container's props, example <ChatContainer component="MockChat" />.
    //    It'll be available in the Container as this.props._component
    // 2. Via constructor of the component that extends this Container.
    //    It'll be available as this.childComponent
    // this.props is checked first to enable dependency injection
    var ChildComponent = this.props._component || this.childComponent;
    var children = this.options.disablePropsChildren ? null : this.props.children;
    return _react2["default"].createElement(ChildComponent, _extends({}, this.getChildProps(this.props), { children: children }));
  };

  return Container;
}(_react2["default"].Component);

exports["default"] = Container;
